src/array_like.rs

```diff
@@ -0,0 +1,111 @@
+//! TODO
+
+use crate::{
+    ndarray::{Array, Array1, ArrayView, Axis, Dimension},
+    Element, PyReadonlyArray,
+};
+use ndarray::{Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+use pyo3::{exceptions::PyValueError, FromPyObject, PyAny, PyResult};
+
+/// TODO
+#[derive(Debug)]
+pub struct PyArrayLike<'py, T, D>(ArrayLike<'py, T, D>)
+where
+    T: Element,
+    D: Dimension;
+
+#[derive(Debug)]
+enum ArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    Reference(PyReadonlyArray<'py, T, D>),
+    Converted(Array<T, D>),
+}
+
+impl<'py, T, D> PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    /// TODO
+    pub fn into_owned_array(self) -> Array<T, D> {
+        match self.0 {
+            ArrayLike::Converted(x) => x,
```

Comment:
I generally prefer using e.g. `Self::Converted` instead of `ArrayLike::Converted` when implementing parts of `ArrayLike`.